File: answers/other/answers/lintcode/Minimum Path Sum.cpp
class Solution {
public:
    /**
     * @param grid: a list of lists of integers.
     * @return: An integer, minimizes the sum of all numbers along its path
     */
    int minPathSum(vector<vector<int> > &grid) {
        // write your code here
		vector<vector<int> > dp(grid.size(), vector<int> (grid[0].size(), INT_MAX));
		for (size_t i = 0; i < grid.size(); ++ i) {
			for (size_t j = 0; j < grid[i].size(); ++ j) {
				dp[i][j] = grid[i][j] + ((i > 0 || j > 0) ? min((i > 0 ? dp[i - 1][j] : INT_MAX), (j > 0 ? dp[i][j - 1] : INT_MAX)) : 0);
			}
		}
		return dp.back().back();
    }
};

File: answers/leetcode/Minimum Path Sum/Minimum Path Sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        List<List<Integer>> sums = new ArrayList<List<Integer>>();
        for (int i = 0; i < grid.length; i++) {
            sums.add(new ArrayList<Integer>());
            for (int j = 0; j < grid[i].length; j++) {
                int minPath = Integer.MAX_VALUE;
                if (i > 0) {
                    minPath = Math.min(minPath, sums.get(i - 1).get(j));
                }
                if (j > 0) {
                    minPath = Math.min(minPath, sums.get(i).get(j - 1));
                }
                if (minPath == Integer.MAX_VALUE) {
                    minPath = 0;
                }
                sums.get(i).add(minPath + grid[i][j]);
            }
        }
        return sums.get(grid.length - 1).get(grid[0].length - 1);
    }
}